File: hw2/q2.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <sstream>

class CStream {
    public:
        std::string fname;
        int line_num;
        int char_pos;
        std::string content;
        int content_pos;

        // Constructor
        CStream(std::string file_name) {
            fname = file_name;
            line_num = -1;
            char_pos = -1;

            // Source: https://stackoverflow.com/questions/2602013/read-whole-ascii-file-into-c-stdstring
            std::ifstream t(fname);
            std::stringstream buffer;
            buffer << t.rdbuf();
            content = buffer.str();
            content_pos = -1;
        }

        // Return true if have more contents to read in file
        bool more_available() {
            // if current position not last character
            if (content_pos == content.length() - 1)
                return false;
            return true;
        }

        // Return character at current position in file
        char get_cur_char() {
            return content[content_pos];
        }

        // Get next character in file and move position
        char get_next_char() {
            if (line_num == -1) { // check initial value of line_num
                line_num++;
            }
            char_pos++;
            content_pos++;
            char ch = ' ';
            while (1) {
                ch = content[content_pos];
                if (ch == '\n') {
                    char_pos = 0; // reset char position to 0 when on new line
                    line_num++;
                    content_pos++;
                } else {
                    return ch;
                }
            }
        }

        // Get next character in file without changing position in file
        char peek_next_char() {
            // peek_ahead_char(0) is same as peek_next_char()
            return peek_ahead_char(0);
        }

        // Get kth character ahead in file without changing position
        char peek_ahead_char(int k) {
            int count = 0; // keep track of non-newline chars
            int i = 1; // keep track of newline chars; default is 1 to make k=0 next char
            char ch = ' ';
            while (count <= k) {
                ch = content[content_pos + count + i]; // check character ahead
                if (ch == '\n') {
                    i++;
                } else {
                    count++;
                }
            }
            return ch;
        }
};

int main() {
    std::cout << std::boolalpha;

    CStream f = CStream("./q2_test1.txt");
    std::cout << "TEST 1" << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.peek_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.peek_ahead_char(4) << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_cur_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.more_available() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;

    f = CStream("./q2_test2.txt");
    std::cout << "\nTEST 2" << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.peek_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.peek_ahead_char(7) << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.peek_ahead_char(6) << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_cur_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.more_available() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;

    f = CStream("./q2_test3.txt");
    std::cout << "\nTEST 3" << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.more_available() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.more_available() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;

    f = CStream("./q2_test4.txt");
    std::cout << "\nTEST 4" << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.peek_ahead_char(0) << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.peek_ahead_char(4) << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.get_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.peek_next_char() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.peek_ahead_char(3) << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;
    std::cout << f.more_available() << std::endl;
    std::cout << "line_num = " << f.line_num << ", char_pos = " << f.char_pos << std::endl;

    return 0;
}

File: hw2/q2.py
class CStream:
    def __init__(self, fname):
        ''' Initializer '''
        self.fname = fname
        self.line_num = -1
        self.char_pos = -1
        self.content = ""
        with open(fname, 'r') as fp:
            self.content = fp.read() # read entire file into string
        self.content_pos = -1 # current index in string

    def more_available(self):
        ''' Return true if have more contents to read in file '''
        if self.content_pos == len(self.content) - 1: # if current position not last character
            return False
        return True

    def get_cur_char(self):
        ''' Return character at current position in file '''
        return self.content[self.content_pos] # return character at current position
    
    def get_next_char(self):
        ''' Get next character in file and move position '''
        if self.line_num == -1: # check initial value of line_num
            self.line_num += 1
        self.char_pos += 1
        self.content_pos += 1
        while True:
            ch = self.content[self.content_pos]
            if ch == '\n':
                self.char_pos = 0 # reset char position to 0 when on new line
                self.line_num += 1
                self.content_pos += 1
            else:
                return ch

    def peek_next_char(self):
        ''' Get next character in file without changing position in file '''
        return self.peek_ahead_char(0) # peek_ahead_char(0) is same as peek_next_char()

    def peek_ahead_char(self, k):
        ''' Get kth character ahead in file without changing position '''
        count = 0 # keep track of non-newline char
        i = 1 # keep track of newline char; default is 1 to make k=0 next char
        ch = ''
        while count <= k:
            ch = self.content[self.content_pos + count + i] # check character ahead
            if ch == '\n':
                i += 1
            else:
                count += 1
        return ch


if __name__ == '__main__':
    print("TEST 1")
    f = CStream('./q2_test1.txt')
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.peek_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.peek_ahead_char(4))
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_cur_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.more_available())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")

    print("\nTEST 2")
    f = CStream('./q2_test2.txt')
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.peek_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.peek_ahead_char(7))
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.peek_ahead_char(6))
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_cur_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.more_available())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")

    print("\nTEST 3")
    f = CStream('./q2_test3.txt')
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.more_available())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.more_available())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")

    print("\nTEST 4")
    f = CStream('./q2_test4.txt')
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.peek_ahead_char(0))
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.peek_ahead_char(4))
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.get_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.peek_next_char())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.peek_ahead_char(3))
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
    print(f.more_available())
    print(f"line_num = {f.line_num}, char_pos = {f.char_pos}")
